PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/utils/tlp.cpp

```diff
@@ -1,15 +1,327 @@
 #include "tlp.h"
+
 int TLP::getTotalLength() {
-	int OHCLength = header->OHCVector.size()*4;
+	int OHCLength = header->OHCVector.size() * 4;
 	int header_length = header->nonBase->headerSizeInBytes;
 	int payload_length = header->lengthInDoubleWord * 4;
 	return (OHCLength + header_length + payload_length);
 }
 
 boost::dynamic_bitset<> TLP::getBitRep() {
-    boost::dynamic_bitset<> result(getTotalLength()*8);
-    result |=  (dataPayload) ;
-    result |= ((boost::dynamic_bitset<>((header->getBitRep())) << getTotalLength() ));
-    
-    return result;
+	boost::dynamic_bitset<> result(getTotalLength() * 8);
+	result |= (dataPayload);
+	result |= ((boost::dynamic_bitset<>((header->getBitRep())) << getTotalLength()));
+
+	return result;
+}
+
+/**
+ * @brief Creates a MemRead32 TLP with the given parameters.
+ *
+ * @param requesterId The requester ID of the TLP.
+ * @param tag The tag field of the TLP header.
+ * @param address The 32-bit address to read from.
+ * @param firstDWBE An array of four integers indicating which bytes are enabled in the first data word (DW).
+ * @param lastDWBE An array of four integers indicating which bytes are enabled in the last DW.
+ * @return A TLP object representing the MemRead32 TLP.
+ */
+TLP TLP::createMemRead32Tlp(int requesterId, int tag, int address, std::bitset<4>  firstDWBE, std::bitset<4> lastDWBE) {
+	TLP memRead32Tlp;
+	memRead32Tlp.header->OHCVector.push_back(new OHCA1(firstDWBE, lastDWBE));
+	memRead32Tlp.header->TLPtype = TLPType::MemRead32;
+	memRead32Tlp.header->lengthInDoubleWord = 0;
+
+		memRead32Tlp.header->nonBase = new AddressRouting32Bit(requesterId, tag, address);
+
+	// clears out the data payload
+	memRead32Tlp.dataPayload.reset();
+	memRead32Tlp.creditConsumedType = Dllp::CreditType::NP;
+	memRead32Tlp.headerConsumption = 1;
+	memRead32Tlp.dataConsumption = 0;
+
+	return memRead32Tlp;
+}
+
+/**
+ * @brief Creates a MemWrite32 TLP with the given parameters.
+ *
+ * @param dataPayloadLength The length of the data payload in bytes.
+ * @param dataPayload The data payload as a string of bits.
+ * @param requesterId The requester ID of the TLP.
+ * @param tag The tag field of the TLP header.
+ * @param address The 32-bit address to write to.
+ * @param firstDWBE An array of four integers indicating which bytes are enabled in the first data word (DW).
+ * @param lastDWBE An array of four integers indicating which bytes are enabled in the last DW.
+ * @return A TLP object representing the MemWrite32 TLP.
+ */
+TLP TLP::createMemWrite32Tlp(int dataPayloadLength, std::string dataPayload, int requesterId, int tag, int address, std::bitset<4>  firstDWBE, std::bitset<4>lastDWBE) {
+	TLP memWrite32Tlp;
+	memWrite32Tlp.header->OHCVector.push_back(new OHCA1(firstDWBE, lastDWBE));
+	memWrite32Tlp.header->TLPtype = TLPType::MemWrite32;
+	memWrite32Tlp.header->lengthInDoubleWord = dataPayloadLength;
```

Comment:
I think there's a confusion here about the meaning of lengthInDoubleWord. If I remember correctly, I think it represented the total header length (including OHCs) in DW.

Reply:
Nope